pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java

```diff
@@ -115,66 +82,46 @@ public boolean isMultiBatches() {
         return true;
     }
 
-    private ProducerImpl.OpSendMsg createOpSendMsg(KeyedBatch keyedBatch) throws IOException {
-        ByteBuf encryptedPayload = producer.encryptMessage(keyedBatch.messageMetadata,
-                keyedBatch.getCompressedBatchMetadataAndPayload());
-        if (encryptedPayload.readableBytes() > ClientCnx.getMaxMessageSize()) {
-            keyedBatch.discard(new PulsarClientException.InvalidMessageException(
-                    "Message size is bigger than " + ClientCnx.getMaxMessageSize() + " bytes"));
-            return null;
-        }
-
-        final int numMessagesInBatch = keyedBatch.messages.size();
-        long currentBatchSizeBytes = 0;
-        for (MessageImpl<?> message : keyedBatch.messages) {
-            currentBatchSizeBytes += message.getDataBuffer().readableBytes();
-        }
-        keyedBatch.messageMetadata.setNumMessagesInBatch(numMessagesInBatch);
-        if (currentTxnidMostBits != -1) {
-            keyedBatch.messageMetadata.setTxnidMostBits(currentTxnidMostBits);
-        }
-        if (currentTxnidLeastBits != -1) {
-            keyedBatch.messageMetadata.setTxnidLeastBits(currentTxnidLeastBits);
-        }
-        ByteBufPair cmd = producer.sendMessage(producer.producerId, keyedBatch.sequenceId, numMessagesInBatch,
-                keyedBatch.messageMetadata, encryptedPayload);
-
-        ProducerImpl.OpSendMsg op = ProducerImpl.OpSendMsg.create(
-                keyedBatch.messages, cmd, keyedBatch.sequenceId, keyedBatch.firstCallback);
-
-        op.setNumMessagesInBatch(numMessagesInBatch);
-        op.setBatchSizeByte(currentBatchSizeBytes);
-        return op;
-    }
-
     @Override
     public List<ProducerImpl.OpSendMsg> createOpSendMsgs() throws IOException {
-        List<ProducerImpl.OpSendMsg> result = new ArrayList<>();
-        List<KeyedBatch> list = new ArrayList<>(batches.values());
-        list.sort(((o1, o2) -> ComparisonChain.start()
-                .compare(o1.sequenceId, o2.sequenceId)
-                .result()));
-        for (KeyedBatch keyedBatch : list) {
-            ProducerImpl.OpSendMsg op = createOpSendMsg(keyedBatch);
-            if (op != null) {
-                result.add(op);
+        try {
+            // In key based batching, the sequence ids might not be ordered, for example,
+            // | key | sequence id list |
+            // | :-- | :--------------- |
+            // | A | 0, 3, 4 |
+            // | B | 1, 2 |
+            // The message order should be 1, 2, 0, 3, 4 so that a message with a sequence id <= 4 should be dropped.
+            // However, for a MessageMetadata with both `sequence_id` and `highest_sequence_id` fields, the broker will
+            // expect a strict order so that the batch of key "A" (0, 3, 4) will be dropped.
+            // Therefore, we should update the `sequence_id` field to the highest sequence id and remove the
+            // `highest_sequence_id` field to allow the weak order.
```

Comment:
> It's a pity to see msg-0 is rejected as the duplicated message. It's true that this PR cannot solve this problem as well.

In addition to [my comment](https://github.com/apache/pulsar/pull/15413#issuecomment-1115325307), it's an issue of the sequence ids cannot be strictly ordered in key based batching. They could only be ordered per key.

Therefore, here we noted that key based batching allows the sequence id order like 1, 2, 0, 3, 4 if 0, 3, 4 are messages of another key. @poorbarcode

Reply:
> What this PR wants to solve is the **resend** case of an existing message. For example, 4 messages might be split into two groups in key based batching:
> 
> * A: 0, 3 (i.e. messages whose keys are all "A", and the sequence ids are 0 and 3)
> * B: 1, 2
> For the 1st case, we can verify it easily with the unit test.

Thank you for your explanation, I see what you mean.  I still have a question:

```java
public void test() throws Exception {
        final String topic = "persistent://my-property/my-ns/test";
        admin.namespaces().setDeduplicationStatus("my-property/my-ns", true);

        final Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
                .topic(topic)
                .batcherBuilder(BatcherBuilder.KEY_BASED)
                .create();
        final List<TypedMessageBuilder<String>> messagesToSend = new ArrayList<>();
        messagesToSend.add(producer.newMessage().value("msg-0").key("A").sequenceId(0));
        messagesToSend.add(producer.newMessage().value("msg-1").key("B").sequenceId(1));
        messagesToSend.add(producer.newMessage().value("msg-3").key("A").sequenceId(3));
        final List<CompletableFuture<MessageId>> futures = new ArrayList<>();
        messagesToSend.forEach(msg -> futures.add(msg.sendAsync()));
        producer.flush(); 
        // didn't receive the ack receipt of batch B,  it might resend batch B.
        ids.add(messagesToSend.get(1).send());
 }
```
If the client didn't receive the ack receipt of batch B,  it might resend batch B.

-- | Before this patch | After this patch
-- | -- | --
msg-1 | accepted | rejected

This patch changed this implements.  

Anyway, I think it's an good catch !

Reply:
> If the client didn't receive the ack receipt of batch B, it might resend batch B.

After this patch the send order will be batch B, batch A as they are sorted by the highest sequence id (1 < 3).

If batch A was not persisted, it would be OK. Otherwise, since batch B was sent before batch A, it means batch A has been persisted as well. So rejecting batch B is reasonable.

Reply:
> After this patch the send order will be batch B, batch A as they are sorted by the highest sequence id (1 < 3).

If batch B has [msg-1, msg-4], then client didn't receive the ack receipt of batch B, it might resend batch B. like this

```
ids.add(messagesToSend.get(1).send());
ids.add(messagesToSend.get(4).send());
```

after this patch, msg-1 will be rejected. msg-4 will be accepted.

Reply:
> msg-1 will be rejected

**It's a correct behavior.** This behavior could be incorrect only when **batch A has been persisted while batch B is not persisted**.  However, batch B must be persisted before batch A.

Reply:
```java
batches.values().forEach(batchMessageContainer -> {
                batchMessageContainer.setLowestSequenceId(batchMessageContainer.getHighestSequenceId());
                batchMessageContainer.setHighestSequenceId(-1L);
            });
            return batches.values().stream().sorted((o1, o2) ->
                    (int) (o1.getLowestSequenceId() - o2.getLowestSequenceId())
            )
```

If batch A has [msg-0, msg-3] and batch B has [msg-1, msg-4],  batch B will persistent after batch A. 

So problems can still happen

Reply:
Yeah. It's right as I've already explained.